aptos-move/framework/aptos-framework/sources/coin.move

```diff
@@ -617,6 +653,207 @@ module aptos_framework::coin {
     }
 
+    /// Initialize fungible asset migration.
+    public entry fun initialize_fungible_asset_migration(aptos_framework: &signer) {
+        system_addresses::assert_aptos_framework(aptos_framework);
+
+        move_to(aptos_framework, FungibleAssetMigration {
+            coins_to_fungible_assets: smart_table::new(),
+            fungible_assets_to_coins: smart_table::new(),
+            enable_default_fungible_assets_in_transfers: false,
+        });
+    }
+
+    /// Create a fungible asset version of an existing Coin. This can only be called by
+    /// the signer corresponding to the address where CoinType is defined/deployed.
+    /// Icon and project uris are new fields in Fungible Asset that don't exist in Coin
+    /// and can be provided or left empty at the caller's discretion.
+    public entry fun initialize_fungible_asset<CoinType>(
+        creator: &signer,
+        maximum_supply: Option<u128>,
+        icon_uri: String,
+        project_uri: String,
+    ) acquires CoinInfo, FungibleAssetMigration {
+        assert!(coin_address<CoinType>() == signer::address_of(creator), EFUNGIBLE_ASSET_MIGRATION_UNAUTHORIZED);
+
+        // if supply is tracked in coin v1, then coin v2 should either have unlimited supply,
+        // or a maximum supply that is equal or greater than the current supply;
+        // if supply is not tracked in coin v1, then coin v2 should have unlimited supply.
+        {
+            let supply_v1 = supply<CoinType>();
+            if (option::is_some(&supply_v1)) {
+                assert!(option::is_none(&maximum_supply) || *option::borrow(&maximum_supply) >= *option::borrow(&supply_v1), EFUNGIBLE_ASSET_MIGRATION_INVALID_MAXIMUM_SUPPLY);
+            } else {
+                assert!(option::is_none(&maximum_supply), EFUNGIBLE_ASSET_MIGRATION_INVALID_MAXIMUM_SUPPLY);
+            }
+        };
+
+        let coin_type = type_info::type_of<CoinType>();
+        let migration = borrow_global_mut<FungibleAssetMigration>(@aptos_framework);
+        assert!(!smart_table::contains(&migration.coins_to_fungible_assets, coin_type), EFUNGIBLE_ASSET_MIGRATION_COINTYPE_INITIALIZED);
+
+        let constructor_ref = object::create_named_object(&create_signer(@aptos_framework), *string::bytes(&type_info::type_name<CoinType>()));
+        primary_fungible_store::create_primary_store_enabled_fungible_asset(
+            &constructor_ref,
+            maximum_supply,
+            name<CoinType>(),
+            symbol<CoinType>(),
+            decimals<CoinType>(),
+            icon_uri,
+            project_uri
+        );
+
+        let metadata = object::object_from_constructor_ref<Metadata>(&constructor_ref);
+        smart_table::add(&mut migration.coins_to_fungible_assets, coin_type, metadata);
+        smart_table::add(&mut migration.fungible_assets_to_coins, metadata, coin_type);
+
+        let signer = object::generate_signer(&constructor_ref);
+        move_to(&signer, FungibleAssetAdmin<CoinType> {
```

Comment:
&signer looks weird as the var name collides with the type. How about fungible_asset_signer?